### PR TITLE
Add malli.util/merge call to route data merge

### LINKED ISSUE #422: malli parameters coercion merging broken
```clj
  (do
    (require '[reitit.ring :as ring] '[reitit.ring.coercion :as coercion] 'reitit.coercion.malli)
    (ring/router ["/{a}" {:parameters {:path [:map [:a string?]]}}
                  ["/{b}" {:parameters {:path [:map [:b string?]]}}]]
                 {:data {:coercion reitit.coercion.malli/coercion
                         :middleware [coercion/coerce-request-middleware]}}))
```

```

  Show: Project-Only All 
  Hide: Clojure Java REPL Tooling Duplicates  (11 frames hidden)

1. Unhandled clojure.lang.ExceptionInfo
   nth not supported on this type: Keyword
   #:reitit.exception{:cause #error {
    :cause "nth not supported on this type: Keyword"
    :via
    [{:type java.lang.UnsupportedOperationException
      :message "nth not supported on this type: Keyword"
      :at [clojure.lang.RT nthFrom "RT.java" 991]}]
    :trace
    [[clojure.lang.RT nthFrom "RT.java" 991]
     [clojure.lang.RT nth "RT.java" 940]
     [malli.core$_parse_entry_syntax$_parse__13937 invoke "core.cljc" 270]
     [malli.core$_parse_entry_syntax$fn__13946 invoke "core.cljc" 273]
     [clojure.core$mapv$fn__8445 invoke "core.clj" 6912]
     [clojure.lang.ArrayChunk reduce "ArrayChunk.java" 63]
     [clojure.core.protocols$fn__8154 invokeStatic "protocols.clj" 136]
     [clojure.core.protocols$fn__8154 invoke "protocols.clj" 124]
     [clojure.core.protocols$fn__8114$G__8109__8123 invoke "protocols.clj" 19]
     [clojure.core.protocols$seq_reduce invokeStatic "protocols.clj" 31]
     [clojure.core.protocols$fn__8146 invokeStatic "protocols.clj" 75]
     [clojure.core.protocols$fn__8146 invoke "protocols.clj" 75]
     [clojure.core.protocols$fn__8088$G__8083__8101 invoke "protocols.clj" 13]
     [clojure.core$reduce invokeStatic "core.clj" 6828]
     [clojure.core$mapv invokeStatic "core.clj" 6903]
     [clojure.core$mapv invoke "core.clj" 6903]
     [malli.core$_parse_entry_syntax invokeStatic "core.cljc" 273]
     [malli.core$_parse_entry_syntax invoke "core.cljc" 269]
     [malli.core$_map_schema$reify__13969 _into_schema "core.cljc" 290]
     [malli.core$into_schema invokeStatic "core.cljc" 908]
     [malli.core$into_schema invoke "core.cljc" 903]
     [malli.core$schema invokeStatic "core.cljc" 918]
     [malli.core$schema invoke "core.cljc" 910]
     [malli.core$walk invokeStatic "core.cljc" 959]
     [malli.core$walk invoke "core.cljc" 954]
     [malli.core$walk invokeStatic "core.cljc" 956]
     [malli.core$walk invoke "core.cljc" 954]
     [malli.util$closed_schema invokeStatic "util.cljc" 122]
     [malli.util$closed_schema invoke "util.cljc" 116]
     [reitit.coercion.malli$create$reify__19044 _request_coercer "malli.cljc" 176]
     [reitit.coercion$request_coercer invokeStatic "coercion.cljc" 80]
     [reitit.coercion$request_coercer invoke "coercion.cljc" 73]
     [reitit.coercion$request_coercers$iter__17680__17684$fn__17685 invoke "coercion.cljc" 125]
     [clojure.lang.LazySeq sval "LazySeq.java" 42]
     [clojure.lang.LazySeq seq "LazySeq.java" 51]
     [clojure.lang.RT seq "RT.java" 535]
     [clojure.core$seq__5402 invokeStatic "core.clj" 137]
     [clojure.core$filter$fn__5893 invoke "core.clj" 2809]
     [clojure.lang.LazySeq sval "LazySeq.java" 42]
     [clojure.lang.LazySeq seq "LazySeq.java" 51]
     [clojure.lang.RT seq "RT.java" 535]
     [clojure.core$seq__5402 invokeStatic "core.clj" 137]
     [clojure.core$seq__5402 invoke "core.clj" 137]
     [reitit.coercion$request_coercers invokeStatic "coercion.cljc" 127]
     [reitit.coercion$request_coercers invoke "coercion.cljc" 122]
     [reitit.ring.coercion$fn__17964 invokeStatic "coercion.cljc" 35]
     [reitit.ring.coercion$fn__17964 invoke "coercion.cljc" 27]
     [reitit.middleware$eval17010$fn__17012 invoke "middleware.cljc" 68]
     [reitit.middleware$eval16880$fn__16881$G__16871__16890 invoke "middleware.cljc" 8]
     [reitit.middleware$eval17002$fn__17003 invoke "middleware.cljc" 51]
     [reitit.middleware$eval16880$fn__16881$G__16871__16890 invoke "middleware.cljc" 8]
     [reitit.middleware$expand_and_transform$fn__17028 invoke "middleware.cljc" 89]
     [clojure.core$keep$fn__8559 invoke "core.clj" 7337]
     [clojure.lang.LazySeq sval "LazySeq.java" 42]
     [clojure.lang.LazySeq seq "LazySeq.java" 51]
     [clojure.lang.RT seq "RT.java" 535]
     [clojure.core$seq__5402 invokeStatic "core.clj" 137]
     [clojure.core.protocols$seq_reduce invokeStatic "protocols.clj" 24]
     [clojure.core.protocols$fn__8146 invokeStatic "protocols.clj" 75]
     [clojure.core.protocols$fn__8146 invoke "protocols.clj" 75]
     [clojure.core.protocols$fn__8088$G__8083__8101 invoke "protocols.clj" 13]
     [clojure.core$reduce invokeStatic "core.clj" 6828]
     [clojure.core$into invokeStatic "core.clj" 6895]
     [clojure.core$into invoke "core.clj" 6887]
     [reitit.middleware$expand_and_transform invokeStatic "middleware.cljc" 90]
     [reitit.middleware$expand_and_transform invoke "middleware.cljc" 85]
     [reitit.middleware$compile_result invokeStatic "middleware.cljc" 117]
     [reitit.middleware$compile_result invoke "middleware.cljc" 112]
     [reitit.ring$compile_result$__GT_endpoint__17332 invoke "ring.cljc" 38]
     [reitit.ring$compile_result$fn__17339 invoke "ring.cljc" 54]
     [clojure.lang.PersistentArrayMap kvreduce "PersistentArrayMap.java" 377]
     [clojure.core$fn__8437 invokeStatic "core.clj" 6845]
     [clojure.core$fn__8437 invoke "core.clj" 6830]
     [clojure.core.protocols$fn__8167$G__8162__8176 invoke "protocols.clj" 175]
     [clojure.core$reduce_kv invokeStatic "core.clj" 6856]
     [clojure.core$reduce_kv invoke "core.clj" 6847]
     [reitit.ring$compile_result invokeStatic "ring.cljc" 51]
     [reitit.ring$compile_result invoke "ring.cljc" 32]
     [reitit.impl$compile_route invokeStatic "impl.cljc" 115]
     [reitit.impl$compile_route invoke "impl.cljc" 114]
     [reitit.impl$compile_routes$fn__16383 invoke "impl.cljc" 118]
     [clojure.core$keep$fn__8559 invoke "core.clj" 7337]
     [clojure.lang.LazySeq sval "LazySeq.java" 42]
     [clojure.lang.LazySeq seq "LazySeq.java" 51]
     [clojure.lang.RT seq "RT.java" 535]
     [clojure.core$seq__5402 invokeStatic "core.clj" 137]
     [clojure.core.protocols$seq_reduce invokeStatic "protocols.clj" 24]
     [clojure.core.protocols$fn__8146 invokeStatic "protocols.clj" 75]
     [clojure.core.protocols$fn__8146 invoke "protocols.clj" 75]
     [clojure.core.protocols$fn__8088$G__8083__8101 invoke "protocols.clj" 13]
     [clojure.core$reduce invokeStatic "core.clj" 6828]
     [clojure.core$into invokeStatic "core.clj" 6895]
     [clojure.core$into invoke "core.clj" 6887]
     [reitit.impl$compile_routes invokeStatic "impl.cljc" 118]
     [reitit.impl$compile_routes invoke "impl.cljc" 117]
     [reitit.core$router invokeStatic "core.cljc" 372]
     [reitit.core$router invoke "core.cljc" 345]
     [reitit.ring$router invokeStatic "ring.cljc" 106]
     [reitit.ring$router invoke "ring.cljc" 77]
     [phoenix.http$eval23803 invokeStatic "NO_SOURCE_FILE" 70]
     [phoenix.http$eval23803 invoke "NO_SOURCE_FILE" 70]
     [clojure.lang.Compiler eval "Compiler.java" 7177]
     [clojure.lang.Compiler eval "Compiler.java" 7167]
     [clojure.lang.Compiler eval "Compiler.java" 7132]
     [clojure.core$eval invokeStatic "core.clj" 3214]
     [clojure.core$eval invoke "core.clj" 3210]
     [clojure.main$repl$read_eval_print__9086$fn__9089 invoke "main.clj" 437]
     [clojure.main$repl$read_eval_print__9086 invoke "main.clj" 437]
     [clojure.main$repl$fn__9095 invoke "main.clj" 458]
     [clojure.main$repl invokeStatic "main.clj" 458]
     [clojure.main$repl doInvoke "main.clj" 368]
     [clojure.lang.RestFn invoke "RestFn.java" 1523]
     [nrepl.middleware.interruptible_eval$evaluate invokeStatic "interruptible_eval.clj" 79]
     [nrepl.middleware.interruptible_eval$evaluate invoke "interruptible_eval.clj" 55]
     [nrepl.middleware.interruptible_eval$interruptible_eval$fn__935$fn__939 invoke "interruptible_eval.clj" 142]
     [clojure.lang.AFn run "AFn.java" 22]
     [nrepl.middleware.session$session_exec$main_loop__1036$fn__1040 invoke "session.clj" 171]
     [nrepl.middleware.session$session_exec$main_loop__1036 invoke "session.clj" 170]
     [clojure.lang.AFn run "AFn.java" 22]
     [java.lang.Thread run "Thread.java" 834]]}}
            exception.cljc:   19  reitit.exception$exception/invokeStatic
            exception.cljc:   15  reitit.exception$exception/invoke
                 core.cljc:  395  reitit.core$router/invokeStatic
                 core.cljc:  345  reitit.core$router/invoke
                 ring.cljc:  106  reitit.ring$router/invokeStatic
                 ring.cljc:   77  reitit.ring$router/invoke
                      REPL:   70  phoenix.http/eval23803
                      REPL:   70  phoenix.http/eval23803
             Compiler.java: 7177  clojure.lang.Compiler/eval
             Compiler.java: 7167  clojure.lang.Compiler/eval
             Compiler.java: 7132  clojure.lang.Compiler/eval
                  core.clj: 3214  clojure.core/eval
                  core.clj: 3210  clojure.core/eval
                  main.clj:  437  clojure.main/repl/read-eval-print/fn
                  main.clj:  437  clojure.main/repl/read-eval-print
                  main.clj:  458  clojure.main/repl/fn
                  main.clj:  458  clojure.main/repl
                  main.clj:  368  clojure.main/repl
               RestFn.java: 1523  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   79  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:   55  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:  142  nrepl.middleware.interruptible-eval/interruptible-eval/fn/fn
                  AFn.java:   22  clojure.lang.AFn/run
               session.clj:  171  nrepl.middleware.session/session-exec/main-loop/fn
               session.clj:  170  nrepl.middleware.session/session-exec/main-loop
                  AFn.java:   22  clojure.lang.AFn/run
               Thread.java:  834  java.lang.Thread/run

```

### PR DESCRIPTION
**Pull request by Deraen**
POC, need to figure out API where Malli coercion can extend the merge functionality, as malli.core can't depend on malli call directly, and this must be enabled only when using malli coercion anyway.

Fixes #422

### CODE REVIEW

**Comment by Deraen** [comment]
- What's the extension point, Coercion protocol maybe? Or is there something else that should be able to extend this?
- Coercion is itself set in the route-data, so we would need to find out the coercion value for route before merging other values?
- Should the API allow to control any paths, or just :parameters data?

**Comment by Deraen** [comment]
Maybe a simple solution:

`router` level options to control route-data merging, provide a function in Malli namespace that can be used for this option. This would need to be set in addition to coercer.

This doesn't allow handling merges if some parts of the route tree use Spec and others Malli etc.

**Comment by ikitommi** [comment]
Maybe the Coercion(?) protocol should have a method for merging parameters?

**Comment by Deraen** [comment]
Coercion value is itself set in the route data, it is hard to get the used coercion before we have merged the route data.

Also, what if the top-level uses Reitit coercion and some routes use Spec coercion, which coercion is used to merge the different levels of route data for those routes?

**Comment by Deraen** [comment]
I have a version where Coercion implementations now implement RouteDataMerge protocol which is used to collect route-data parts into one map.

The problems:
- Route data merge is currently tied to :coercion route data, which will currently affect :parameters route data.
- The effective :coercion value needs to be retrieved separately before the merge
- There could be need to control other route data keys than just :parameters

I think better solution would be to just add some kind of declaration or function to router options, to control how keys are merged. This will allow controlling other keys, and doesn't tie this together with coercion values.

File: modules/reitit-core/src/reitit/impl.cljc [hunk 1]
```diff
@@ -60,14 +63,31 @@
...
+  (let [;; Find last the effective :coercion value
+        ;; for the route, and then use the cocercion
+        ;; instance for route data merge implementation.
+        ;; TODO: other options/keys should also be able to control
+        ;; merge?
+        coercion (->> x
+                      reverse
+                      (some (fn [[k v]]
+                              (when (= :coercion k)
+                                v))))
```
Comment (Deraen) [inline modules/reitit-core/src/reitit/impl.cljc:79]: This is the messy part to find effective coercion value.

**Comment by ikitommi** [comment]
This is one way to resolve this, also not happy how coercion spills into core ns. Also, if you have different coercion impls at different parts of the route tree, how does the merge work? doesn't? e.g.

```clj
["/root" {:coercion spec, :parameters {:query {:x int?}}}
 ["/schema" {:coercion schema, :parameters {:query {:y s/Int}}}
  |"/malli" {:coercion malli, :parameters {:query [:map [:z :int]]}}]]]
```
I guess it breaks?

Questions:
1. do we need to support multiple coercion impls in one route-tree? => why not
2. do we have to have coherent coercion impl in a route tree + parents => YES

One way to have a clean separation of concerns would be to define a generic parameter merge strategy per parameters path, without coercion. e.g. a `reitit.ring/router` could say that `:parameters` is a special path in which, the values are accumulated into a vector. This way, we have all the schemas separately and could use Coercion impl to merge those in a second sweep.

```clj
;; original
["/api" {:coercion malli, :parameters {:query [:map [:x :int]]}}
 ["/admin" {:parameters {:query [:map [:y :int]]}}
  |"/users" {:parameters {:query [:map [:z :int]]}}]]]

; after path & route data merging
["/api/admin/users {:coercion malli, :parameters [{:query [:map [:x :int]]} {:query [:map [:y :int]]} {:query [:map [:z :int]]}]}]

;; after coercion has merged the params
["/api/admin/users {:coercion malli, :parameters [{:query [:map [:x :int] [:y :int] [:z :int]]}]
```

what do you think?

**Comment by ikitommi** [comment]
... the actual impl for doing the accumulation could be done either a new Protocol or just meta-data. We might need to fork `meta-merge` for this.

```clj
;; via meta-data
(meta-merge/merge {:parameters ^:accumulate []} {:parameters {:query [:map [:x :int]]}})
; => {:parameters [{:query [:map [:x :int]]}])

;; via protocol (and records)
(meta-merge/accumulator)
; => #meta.merge.Accumulator{:data []}

(meta-merge/merge {:parameters *1} {:parameters {:query [:map [:x :int]]}})
; => {:parameters #meta.merge.Accumulator{:data [{:query [:map [:x :int]]}]})
```

**Comment by Deraen** [comment]
Replaced by ce06214014499caf86d3abd65884a85e7004b53d

**Pull request closed without merging**